docs/topics/addons.rst

```diff
@@ -0,0 +1,348 @@
+
+.. attribute:: provides
+
+    list of components provided by this add-on (e.g. ``mongodb`` for an
+    extension that provides generic read/write access to a MongoDB database)
+
+The entries in the :attr:`requires` and :attr:`modifies` attributes can be add-on
+names or components from other add-ons' :attr:`provides` attribute. You can
+specify :pep:`440`-style information about required versions. Examples::
+
+    requires = ['httpcache']
+    requires = ['otheraddon >= 2.0', 'yetanotheraddon']
+
+The Python object or module that is pointed to by an add-on path (e.g. given in
+the ``ADDONS`` setting, or given to
+:meth:`~scrapy.addons.AddonManager.add`) does not necessarily have to be an
+add-on. Instead, it can provide an ``_addon`` attribute. This attribute can be
+either an add-on or another add-on path.
+
+
+Add-on base class
+=================
+
+Scrapy comes with a built-in base class for add-ons which provides some
+convenience functionality:
+
+* basic settings can be exported via :meth:`~scrapy.addons.Addon.export_basics`,
+  configurable via :attr:`~scrapy.addons.Addon.basic_settings`.
+* a single component (e.g. an item pipeline or a downloader middleware) can be
+  inserted into Scrapy's settings via
+  :meth:`~scrapy.addons.Addon.export_component`, configurable via
+  :attr:`~scrapy.addons.Addon.component_type`,
+  :attr:`~scrapy.addons.Addon.component_key`,
+  :attr:`~scrapy.addons.Addon.component`, and the ``order`` key in
+  :attr:`~scrapy.addons.Addon.default_config`.
+* the add-on configuration can be exposed into Scrapy's settings via
+  :meth:`~scrapy.addons.Addon.export_config`, configurable via
+  :attr:`~scrapy.addons.Addon.default_config`,
+  :attr:`~scrapy.addons.Addon.config_mapping`, and
+  :attr:`~scrapy.addons.Addon.settings_prefix`.
+
+By default, the base add-on class will expose the add-on configuration into
+Scrapy's settings namespace, in caps and with the add-on name prepended. It is
+easy to write your own functionality while still being able to use the
+convenience functions by overwriting
+:meth:`~scrapy.addons.Addon.update_settings`.
+
+.. module:: scrapy.addons
+   :noindex:
+
+.. autoclass:: Addon
+   :members:
+
+
+Add-on examples
+===============
+
+Set some basic configuration using the :class:`Addon` base class::
+
+    from scrapy.addons import Addon
+
+    class MyAddon(Addon):
+        name = 'myaddon'
+        version = '1.0'
+        component = 'path.to.mypipeline'
+        component_type = 'ITEM_PIPELINES'
+        component_order = 200
+        basic_settings = {
+            'DNSCACHE_ENABLED': False,
+        }
+
+Check dependencies::
+
+    from scrapy.addons import Addon
+
+    class MyAddon(Addon):
+        name = 'myaddon'
+        version = '1.0'
+
+        def update_settings(self, config, settings):
+            try:
+                import boto
+            except ImportError:
+                raise RuntimeError("myaddon requires the boto library")
```

Comment:
I wonder what's the benefit of checking it here, as opposed to checking it in the component's `__init__`. If it's checked in the component, we get the same error behavior when addons system is not used.